snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/utils/DeviceInfoMonitor.java

```diff
@@ -0,0 +1,319 @@
+     *
+     * @param context the android context
+     * @return the representation of the current network connection or null
+     */
+    @Nullable
+    public NetworkInfo getNetworkInfo(@NonNull Context context) {
+        ConnectivityManager cm = (ConnectivityManager)
+                context.getSystemService(Context.CONNECTIVITY_SERVICE);
+
+        NetworkInfo ni = null;
+        try {
+            NetworkInfo maybeNi = cm.getActiveNetworkInfo();
+            if (maybeNi != null && maybeNi.isConnected()) {
+                ni = maybeNi;
+            }
+        } catch (SecurityException e) {
+            Logger.e(TAG, "Security exception getting NetworkInfo: %s", e.toString());
+        }
+        return ni;
+    }
+
+    /**
+     * @param context The Android context
+     * @return Total physical memory on the device in bytes
+     */
+    public long getPhysicalMemory(@NonNull Context context) {
+        ActivityManager.MemoryInfo mi = getMemoryInfo(context);
+        return mi.totalMem;
+    }
+
+    /**
+     * @param context The Android context
+     * @return Currently available system memory in bytes
+     */
+    public long getSystemAvailableMemory(@NonNull Context context) {
+        ActivityManager.MemoryInfo mi = getMemoryInfo(context);
+        return mi.availMem;
+    }
+
+    /**
+     * @param context The Android context
+     * @return A pair containing the current battery state (either "full", "charging", "unplugged" or NULL if unknown) and the battery level (0 to 100)
+     */
+    @Nullable
+    public Pair<String, Integer> getBatteryStateAndLevel(@NonNull Context context) {
+        IntentFilter ifilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
+        Intent batteryIntent = context.registerReceiver(null, ifilter);
+        if (batteryIntent == null) {
+            return null;
+        }
+
+        int status = batteryIntent.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
+        int level = batteryIntent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
+        int scale = batteryIntent.getIntExtra(BatteryManager.EXTRA_SCALE, -1);
+        if (status == -1 || level == -1 || scale == -1) {
+            return null;
+        }
+
+        String batteryState;
+        switch (status) {
+            case BatteryManager.BATTERY_HEALTH_UNKNOWN:
+                batteryState = null;
+                break;
+            case BatteryManager.BATTERY_STATUS_FULL:
+                batteryState = "full";
+                break;
+            case BatteryManager.BATTERY_STATUS_CHARGING:
+                batteryState = "charging";
+                break;
+            default:
+                batteryState = "unplugged";
+                break;
+        }
+
+        int batteryLevel = (int) (level * 100 / (float)scale);
+
+        return new Pair<>(batteryState, batteryLevel);
+    }
+
+    /**
+     * Checked in the user data directory.
+     * @return Currently available storage on disk in bytes
+     */
```

Comment:
This comments are great, but don't feel that you have to add api comments to all the methods.
IMO the javadoc on the public API is already enough. We can get rid of comments in the internal classes even if their methods are public methods. Actually, I would completely exclude them from the Java doc as those classes shouldn't be used by the final users. Though, I think comments are very valuable in cases where the behaviour is hard to guess. In this case, your code is very well written and easy to read!
Btw, I'm just sharing my point of view, so no need to remove the comment in this PR now.

Reply:
I agree that some of the methods are pretty self-explanatory and the comments don't add much. I removed some such comments for the internal classes. I'm not sure whether removing all comments for internal methods is a good idea, sometimes they add useful information for us. It seems to be possible to [exclude packages from Javadoc configuration](https://maven.apache.org/plugins/maven-javadoc-plugin/examples/exclude-package-names.html), so we can also try that.